### PR TITLE
rubicon project adapter

### PR DESCRIPTION
**Pull request by mhc-gh**
Initial submission of Rubicon Project adapter

### CODE REVIEW

**Comment by mhc-gh** [comment]
Any idea why the travis build is failing on a polyfill.js test?  It's not a file i've edited.

**Comment by cramforce** [comment]
Test failure is a frequent flake that we are working on fixing.

File: ads/rubicon.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {writeScript, loadScript, checkData} from '../src/3p';
+//import {nonSensitiveDataPostMessage, listenParent} from '../3p/messaging.js';
```
Comment (cramforce) [inline ads/rubicon.js]: Remove

File: ads/rubicon.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    gptran = true;
+
+    data.targeting['rpfl_' + data.account] =
+      rubicontag.getSlot('c').getAdServerTargeting();
+
+    data.targeting['rpfl_elemid'] = 'c';
+    doubleclick(global, data);
+  };
+
+  loadScript(global, 'https://ads.rubiconproject.com/header/' + data.account + '.js', () => {
```
Comment (cramforce) [inline ads/rubicon.js]: Please use `encodeURIComponent` on the account value.

File: ads/rubicon.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      const slot = rubicontag.defineSlot(data.slot, dimensions, 'c');
+
+      slot.setPosition(data.pos);
+
+      rubicontag.addKW(data.kw);
+      rubicontag.setFPV(data.visitor);
+      rubicontag.setFPI(data.inventory);
+
+      rubicontag.run(gptrun, 1000);
+
```
Comment (cramforce) [inline ads/rubicon.js]: nit: please remove the vertical whitespace

File: ads/rubicon.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  global.rp_account = data.account;
+  global.rp_site = data.site;
+  global.rp_zonesize = data.zone + '-' + data.size;
+  global.rp_adtype = 'js';
+  global.rp_kw = data.kw;
+  global.rp_visitor = data.visitor;
+  global.rp_inventory = data.inventory;
+  global.rp_callback = data.callback;
+  /* eslint-enable */
+  writeScript(global, 'https://ads.rubiconproject.com/ad/' + data.account + '.js');
```
Comment (cramforce) [inline ads/rubicon.js]: Add `encoreURIComponent` here as well. In this case it is an XSS not to do it.

File: ads/rubicon.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+For semantics of configuration, please contact your Rubicon Account Director.
+
+
+#### Ad size
+
+By default the ad size is based on the `width` and `height` attributes of the `amp-ad` tag.
+
+#### Supported parameters
+
```
Comment (cramforce) [inline ads/rubicon.md]: Please add link to your own docs for these.

File: ads/rubicon.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS-IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Rubicon Project
+
+If you want to serve ads via your Ad Server then there is no need to use the adapter when using Smart Tags.  These can simply be served via your Ad Server in the normal fashion.  You simply need to ensure that you are using secure tags (https).
```
Comment (cramforce) [inline ads/rubicon.md]: Add a link to your site.

File: package.json [hunk 7]
```diff
@@ -40,7 +40,7 @@
     "eslint-plugin-google-camelcase": "0.0.2",
     "finalhandler": "0.4.0",
     "fs-extra": "0.26.4",
-    "gulp": "3.9.1",
+    "gulp": "^3.9.1",
```
Comment (cramforce) [inline package.json]: Please remove

File: ads/rubicon.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+- `cookieOptions`
+- `tagForChildDirectedTreatment`
+- `targeting`
+
+##### First Party Data & Keywords
+- `data-kw`
+- `data-visitor`
+- `data-inventory`
+
+
```
Comment (cramforce) [inline ads/rubicon.md]: nit: remove trailing line

**Comment by cramforce** [comment]
Looks good. Some small comments. Please squash commits.

**Comment by mhc-gh** [comment]
Okay, something went horribly wrong while trying to squash commits --- trying to undo

**Pull request closed without merging**